src/transformStateWithClones.js

```diff
@@ -7,7 +7,24 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
-}
+  const result = [];
+
+  let transClone = structuredClone(state);
+
+  for (const action of actions) {
+    if (action.type === 'removeProperties') {
+      for (const key of action.keysToRemove) {
+        delete transClone[key];
+      }
+    } else if (action.type === 'addProperties') {
+      transClone = { ...transClone, ...action.extraData };
```

Comment:
When adding properties, ensure that `action.extraData` is a valid object to merge with `transClone`. Consider adding validation or error handling for unexpected data types.